File: 1-Arrays/Container_With_Most_Water.java
class Container_With_Most_Water {
    public int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int maxArea = 0;
        
        while(left <= right) {
            
            int currentArea = (right - left) * Math.min(height[left], height[right]);
            
            maxArea = Math.max(maxArea, currentArea);
            
            if (height[right] > height[left]) {
                left++;
            } else {
                right--;
            }
        }
        
        return maxArea;
    }
}

// Time complexity: O(N) Single pass. 
// Space Complexity: O(1) Constant space is used. 

File: Python/1_Array/Container_With_Most_Water.py
from ast import List


class Solution:
    def maxWater(self, height: List[int]) -> int:
        L, R, maxArea = 0, len(height), 0;

        while L <= R:

            currentArea = (R - L) * min(height[L], height[R])
            maxArea = max(maxArea, currentArea)
            
            if (height[R] > height[L]) {
                L+=;
            } else {
                R+=1;
            }

        return maxArea
    # t: O(N)
    # s: O(1)
    

        